app/service/auth_svc.py

```diff
@@ -144,47 +177,106 @@ async def is_request_authenticated(self, request):
 
     """ PRIVATE """
 
-    @staticmethod
-    async def _check_credentials(user_map, username, password):
-        user = user_map.get(username)
-        if not user:
-            return False
-        return user.password == password
-
-    async def _ldap_login(self, username, password):
-        server = ldap3.Server(self.ldap_config.get('server'))
-        dn = self.ldap_config.get('dn')
-        user_attr = self.ldap_config.get('user_attr') or 'uid'
-        user_string = '%s=%s,%s' % (user_attr, username, dn)
-
-        try:
-            with ldap3.Connection(server, user=user_string, password=password) as conn:
-                if conn.bind():
-                    if username not in self.user_map:
-                        group = await self._ldap_get_group(conn, dn, username, user_attr)
-                        await self.create_user(username, None, group)
-                    return True
-        except LDAPException:
-            self.log.error('Unable to connect to LDAP server')
+    def _set_primary_login_handler(self):
+        """
+        Returns the login handler for the auth service as specified in the config file.
+        This login handler will take priority for login methods during login_user and redirects during
+        check_permissions. If no login handler was specified in the config file, the default handler will be returned.
+        """
+        login_handler_module_path = self.get_config('auth.login.handler.module')
+        if login_handler_module_path and login_handler_module_path != 'default':
+            try:
+                login_handler = import_module(login_handler_module_path).load_login_handler(self.get_services())
+                if isinstance(login_handler, LoginHandlerInterface):
+                    self.log.info('Setting primary login handler: %s' % login_handler_module_path)
+                    return login_handler
+                else:
+                    self.log.warn('Attempted to set login handler that does not implement LoginHandlerInterface.')
+            except Exception as e:
+                self.log.error('Invalid login handler module: %s' % e)
+
+        self.log.info('Using default login handler.')
+        return self._default_login_handler
+
+    class DefaultLoginHandler(LoginHandlerInterface):
```

Comment:
Why a nested class rather than having it be in its own module?

Reply:
I think at the time I figured it would just stay within the auth service, but I think it would make more sense to put it in a separate module, especially since custom login handlers may want to reference the default login handler for whatever reason.